soc: wch: Deduplicate common files across ch32v socs

miggazElquez opened the pull request:
Moves shared files from multiple ch32v SoCs into a new `common/` directory to eliminate duplication.

I only de-duplicated `pinctrl_soc.h` for the qingke_v4x socs, because the pin controller is not the same between the qingke_v2a (ch32v003) and the qingke_v2c (other ch32v00x).

miggazElquez committed:
soc: wch: Deduplicate common files across ch32v socs

Moves shared files from multiple ch32v SoCs into a new `common/`
directory to eliminate duplication.

Signed-off-by: Miguel Gazquez <miguel.gazquez@bootlin.com>

nzmichaelh commented on soc/wch/ch32v/common/qingke_v4/pinctrl_soc.h:
```diff
@@ -10,7 +10,7 @@
  * @brief Type to hold a pin's pinctrl configuration.
  */
-struct ch32v208_pinctrl_soc_pin {
+struct qingke_v4_pinctrl_soc_pin {
```
I don't think we should do this. The pinmux is a feature of the SOC, not the microarchitecture.

For example, both the CH32V003 and CH32V006 are QingKe V2, but they have different pinmux.

I'm concerned this PR will have the same issues as the previous one where there's lots of rebasing due to differing opinions. Should we work this out in a bug or chat first?

nzmichaelh commented on soc/wch/ch32v/common/qingke_v4/pinctrl_soc.h:
```diff
@@ -10,7 +10,7 @@
  * @brief Type to hold a pin's pinctrl configuration.
  */
-struct ch32v208_pinctrl_soc_pin {
+struct qingke_v4_pinctrl_soc_pin {
```
I don't think we should do this. The pinmux is a feature of the SOC, not the microarchitecture.

For example, both the CH32V003 and CH32V006 are QingKe V2, but they have different pinmux.

I'm concerned this PR will have the same issues as the previous one where there's lots of rebasing due to differing opinions. Should we work this out in a bug or chat first?

miggazElquez replied:
> I don't think we should do this. The pinmux is a feature of the SOC, not the microarchitecture.
> For example, both the CH32V003 and CH32V006 are QingKe V2, but they have different pinmux.

I understand this, that's why I didn't move them. For now all qingke_v4-based soc are part of the ch32v20x/30x series and they all have the same pinmux, but if there is another SoC with a qingke_v4  it would indeed be a problem.
But it would still be good to deduplicate this file no ?

nzmichaelh commented on soc/wch/ch32v/ch32v00x/CMakeLists.txt:
```diff
@@ -3,8 +3,8 @@
 zephyr_sources(
-  soc_irq.S
-  vector.S
+  ../common/qingke_v2x/soc_irq.S
```
I don't think a directory should refer to a different directory. I can't find any examples of this in the other socs.

As soc_irq.S is fairly trivial, shall we just denormalise it across all SOCs?

miggazElquez committed:
soc: wch: rename pinctrl struct for the ch32v303

Rename the struct ch32v303_pinctrl_soc_pin to ch32v30x_pinctrl_soc_pin
to allow the same struct to be used for other SoC of the series.

Signed-off-by: Miguel Gazquez <miguel.gazquez@bootlin.com>

miggazElquez commented:
I rebased to include the ch32v303. For the pinctrl file, I moved it back to the folder for each SoC series, but renamed the one in the qingke_v4f folder to 30x from 303, that way it can be re-used by other 30x SoC.

Does it seems like a good solution ?

SoftwareArchitekt reviewed (changes requested):
I just see a problem with the vector.s - file for qinke4. Here i already told in last pr (adding ch32v303), that the 2 assembly lines not work on my ch32v307. Means, we need here a change.
